File: platforms/leetcode/search-a-2d-matrix/solutions/solution.go
func searchMatrix(matrix [][]int, target int) bool {
	m, n := len(matrix), len(matrix[0])

	ridx := -1
	l, r := 0, m-1
	for l <= r {
		mid := (l + r) / 2
		mv := matrix[mid][0]
		if mv == target {
			return true
		}
		if mv > target {
			r = mid - 1
		} else {
			ridx = mid
			l = mid + 1
		}
	}

	if ridx < 0 {
		return false
	}

	l, r = 0, n-1
	for l <= r {
		mid := (l + r) / 2
		mv := matrix[ridx][mid]
		if mv == target {
			return true
		}
		if mv > target {
			r = mid - 1
		} else {
			l = mid + 1
		}
	}

	return false
}

File: platforms/leetcode/search-a-2d-matrix/solutions/solution.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m, n = len(matrix), len(matrix[0])

        ridx = -1
        l, r = 0, m - 1
        while l <= r:
            mid = (l + r) // 2
            mv = matrix[mid][0]
            if mv == target:
                return True
            if mv < target:
                l = mid + 1
                ridx = mid
            else:
                r = mid - 1

        if ridx == -1:
            return False

        l, r = 0, n - 1
        while l <= r:
            mid = (l + r) // 2
            mv = matrix[ridx][mid]
            if mv == target:
                return True
            if mv < target:
                l = mid + 1
            else:
                r = mid - 1

        return False


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m, n = len(matrix), len(matrix[0])

        l, r = 0, m * n - 1

        while l <= r:
            mid = (l + r) // 2
            val = matrix[mid // n][mid % n]

            if val == target:
                return True
            elif val < target:
                l = mid + 1
            else:
                r = mid - 1

        return False
